### PR TITLE
Diagnostic Verifier

### PR DESCRIPTION
**Pull request by harlanhaskins**
This is a first pass at a Swift-style diagnostic verifier for Silt code.

Diagnostics are verified by comparing the full set of diagnostics produced during a compile with a set of expectations specified in comments in the same file.

For example, the following would verify:

```haskell
module Foo Bar where -- expected-error {{unexpected token 'Bar'}}
id x = x
```

But if the diagnostic was not emitted during compile, this would give its own diagnostic:

```
parse-errors.silt:3:15: error: error "unexpected token 'Bar'" never produced
```

### CODE REVIEW

**Comment by harlanhaskins** [comment]
Gonna merge this after we add proper parser diagnostics. Then we can start adding test cases.

**Comment by harlanhaskins** [comment]
Now that we've merged #32, I can go ahead and add some diagnostic verification tests :tada:

File: Sources/silt/main.swift
Comment (CodaFi) [inline Sources/silt/main.swift]: Nit: Could this just be called `verify`?

File: Sources/Lithosphere/DiagnosticConsumer.swift
Comment (CodaFi) [inline Sources/Lithosphere/DiagnosticConsumer.swift]: Would it be overkill to drop this and hand back an `NSNotificationCenter`-style token that has to be used for de-registering consumers?

File: Sources/Drill/DiagnosticVerifier.swift
Comment (CodaFi) [inline Sources/Drill/DiagnosticVerifier.swift]: Nit: Guard on the number of ranges being valid too.

**Comment by CodaFi** [comment]
How does this handle diagnostics that are emitted without a valid source location?

File: Sources/silt/main.swift
Comment (harlanhaskins) [reply Sources/silt/main.swift]: I possibly want to add a verify step for different layers -- maybe this could be `--verify parse`?

File: Sources/Lithosphere/DiagnosticConsumer.swift
Comment (harlanhaskins) [reply Sources/Lithosphere/DiagnosticConsumer.swift]: I suppose I could do that, yeah!

File: Sources/Drill/DiagnosticVerifier.swift
Comment (harlanhaskins) [reply Sources/Drill/DiagnosticVerifier.swift]: 👍

**Comment by harlanhaskins** [comment]
@CodaFi Right now it falls over (you can't make an expectation except within a source file anyway...)

**Comment by CodaFi** [comment]
One last thing: The current architecture doesn't handle partial matches or regular expressions in the expectation sheds.

**Comment by harlanhaskins** [comment]
Yes, you’re right. I should actually look at how Swift’s works and maybe do a Real Port of the matching logic

**Comment by harlanhaskins** [comment]
Okay, we now have:
- Partial matches:
```haskell
id : x -> x --expected-error{{unexpected token 'id'}}
```
- Regex matches:
```haskell
id : x -> x --expected-error{{unexpected {{\w+}} 'id'}}
```
- Line-relative matches:
```haskell
id : x -> x
--expected-error @-1 {{unexpected {{\w+}} 'id'}}
```

**Comment by harlanhaskins** [comment]
Ping @CodaFi (this is surprisingly good now)

File: Sources/Drill/DiagnosticRegexParser.swift
Comment (CodaFi) [inline Sources/Drill/DiagnosticRegexParser.swift]: NSRegularExpression can do this for you.

Comment (harlanhaskins) [reply Sources/Drill/DiagnosticRegexParser.swift]: Oh thank god, I looked for a method for this but couldn't find anything.

Comment (harlanhaskins) [reply Sources/Drill/DiagnosticRegexParser.swift]: Ah oops, found it.

**Comment by harlanhaskins** [comment]
@CodaFi Check the latest macOS failure -- not sure why FileCheck is spinning forever

File: Sources/silt/main.swift
Comment (CodaFi) [inline Sources/silt/main.swift]: Nit: Needs to be indented a little more

File: Tests/DiagnosticTests/Resources/no-module.silt
Comment (CodaFi) [inline Tests/DiagnosticTests/Resources/no-module.silt:2]: Wow, we need a better diagnostic here (Starter Bug!)

**Review by CodaFi — APPROVED** [review]
One nit but nothing that should hold this up.

This Is Very Good

**Comment by CodaFi** [comment]
Ugh, build is spinning because the edit distance algorithm is awful.  I’ll add heuristics that turn it off for big buffers.

**Comment by CodaFi** [comment]
There you go, a real failure.

File: Tests/DiagnosticTests/DiagnosticTests.swift
Comment (CodaFi) [inline Tests/DiagnosticTests/DiagnosticTests.swift]: This needs to be added to `allTests`

**Comment by harlanhaskins** [comment]
:boat:

**Pull request merged**
